src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/InstalledAndTransitivePackageFeed.cs

```diff
@@ -22,18 +22,57 @@ public InstalledAndTransitivePackageFeed(IEnumerable<PackageCollectionItem> inst
             _transitivePackages = transitivePackages ?? throw new ArgumentNullException(nameof(transitivePackages));
         }
 
+        internal static PackageCollectionItem SelectTransitiveLatestPackage(IGrouping<string, PackageCollectionItem> groupedPackagesById, IEnumerable<PackageCollectionItem> installedPkgs)
+        {
+            if (groupedPackagesById == null)
+            {
+                throw new ArgumentNullException(nameof(groupedPackagesById));
+            }
+
+            if (installedPkgs == null)
+            {
+                throw new ArgumentNullException(nameof(installedPkgs));
+            }
+
+            IEnumerable<PackageCollectionItem> matchedOrigins = groupedPackagesById
+                .Where(x => x.PackageReferences
+                    .Cast<ITransitivePackageReferenceContextInfo>()
+                    .Any(trPr => trPr.TransitiveOrigins.Any(trOrigin => installedPkgs.Contains(trOrigin.Identity))));
+
+            if (matchedOrigins.Any())
+            {
+                // Return the latest transitive package with installed versions
+                return matchedOrigins.OrderByDescending(x => x.Version).First();
+            }
+
+            // Otherwise, fallback on what's availble
+            return groupedPackagesById.OrderByDescending(x => x.Version).First();
+        }
+
         /// <inheritdoc cref="IPackageFeed.ContinueSearchAsync(ContinuationToken, CancellationToken)" />
         public override async Task<SearchResult<IPackageSearchMetadata>> ContinueSearchAsync(ContinuationToken continuationToken, CancellationToken cancellationToken)
         {
             var searchToken = continuationToken as FeedSearchContinuationToken ?? throw new InvalidOperationException(Strings.Exception_InvalidContinuationToken);
             cancellationToken.ThrowIfCancellationRequested();
 
+            PackageCollectionItem[] installedFeedItems = _installedPackages.GetLatest();
+            PackageCollectionItem[] matchesInstalled = PerformLookup(installedFeedItems, searchToken);
+
             // Remove transitive packages from project references
             IEnumerable<PackageCollectionItem> pkgsWithOrigins = _transitivePackages
                 .Where(t => t.PackageReferences.Any(x => x is ITransitivePackageReferenceContextInfo y && y.TransitiveOrigins.Any()));
 
-            IPackageSearchMetadata[] installedItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(_installedPackages, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
-            IPackageSearchMetadata[] transitiveItems = await GetMetadataForPackagesAndSortAsync(PerformLookup(pkgsWithOrigins, searchToken), searchToken.SearchFilter.IncludePrerelease, cancellationToken);
+            // Search all transitive packages, then group by package ID later
```

Comment:
Yes, because we want to prioritize transitive packages with transitive origins in the Installed part.

If I group transitive packages by Id first, then, lookup, I would show package versions whose transitive origin does not match the package shown in Installed packages.

Reply:
Added this as a comment.